packages/playwright-test/src/webServer.ts

```diff
@@ -118,11 +119,32 @@ async function isPortUsed(port: number): Promise<boolean> {
   });
 }
 
-async function waitForSocket(port: number, delay: number, cancellationToken: { canceled: boolean }) {
+async function isURLAvailable(url: URL) {
+  return new Promise<boolean>(resolve => {
+    http.get(url, res => {
+      res.resume();
+      const statusCode = res.statusCode ?? 0;
+      resolve(statusCode >= 200 && statusCode < 300);
+    }).on('error', () => {
+      resolve(false);
+    });
+  });
+}
+
+async function waitFor(waitFn: () => Promise<boolean>, delay: number, cancellationToken: { canceled: boolean }) {
   while (!cancellationToken.canceled) {
-    const connected = await isPortUsed(port);
+    const connected = await waitFn();
     if (connected)
       return;
     await new Promise(x => setTimeout(x, delay));
   }
 }
+
+async function waitForSocketOrURL(port: number, path: string | undefined, delay: number, cancellationToken: { canceled: boolean }) {
+  if (path) {
```

Comment:
Let's check here that exactly one of `port` and `path` is undefined, and the other is not.